### DISCUSSION TITLE
How do I mix shared and static libs

### DISCUSSION THREAD

**Discussion by Phil67ago** [mesonbuild/meson#10067] [General]
Following failed on '-lshared-lib' because it is a shared lib only. 
The rest ( '-lcrypto', '-lssl') is okay because they exist as static and shared libs.
On meson 0.59 how do I mix shared and static libs?
'-static' makes it static.

project('foo', 'cpp')

project_sources   = []
project_includes  = [ include_directories('include/','src/') ]
project_c_args    = [ '-ffunction-sections', '-fPIC' ]    
project_link_args = ['-static', '-Wl,-gc-sections' , '-lshared-lib', '-lcrypto', '-lssl' ] 
subdir('src')

exe=executable( 'foo', 
            project_sources, 
            include_directories : [ project_includes ],
            c_args : project_c_args, 
            link_args : project_link_args,  
          )

**Comment by eli-schwartz** [comment]
This isn't something Meson is doing. The `-static` compiler option says to the compiler that only static libraries are allowed. As a result, the compiler fails when static libraries aren't available.

That option also says to statically link to the platform C standard library (e.g. glibc).

What's your goal here? If you just want to link some of your dependencies statically, but you don't want to build a "fully static" executable, don't use the `-static` compiler option.

You're also manually using `-l` flags for libraries, but Meson doesn't like it when you do this. Instead you should use `dependency('openssl')` or `dependency('libssl')` or `dependency('libcrypto')` (depending on which parts of the OpenSSL project you need to link to) and Meson will figure out the best way to link to it.

This allows you to pass the `static: true` kwarg to the dependency lookup, which forces Meson to detect the ssl/crypto libraries as static libraries only.

Meanwhile `cc.find_library('shared-lib')` will find that other library and allow either static or shared (probably shared) versions.

**Comment by Phil67ago** [comment]
Tested and yes that seems to work!
/Thanks


project('foo', 'cpp')

project_sources   = []
project_includes  = [ include_directories('include/','src/') ]
project_c_args    = [ '-ffunction-sections', '-Os', '-fPIC' ]    
project_link_args = [ '-Wl,-gc-sections' ] 
project_dep       = [ dependency('shared-lib'), dependency('crypto'), dependency('ssl',static:true) ]
subdir('src')

exe=executable( 'foo', 
            project_sources, 
            include_directories : [ project_includes ],
            c_args : project_c_args, 
            dependency : project_dep
            link_args : project_link_args,  
          )

**Reply by eli-schwartz** [reply]
Great, happy to help.

Note there are some other optimizations you can do. Instead of passing `'-Os'` you can use `project(..., default_options: ['optimization=s'])` and I'm pretty sure `-fPIC` is the default already.

The advantage of using default_options is that users can override it if they *really* want, and either way Meson will handle the option correctly across different compilers.
